src/plugins/oidc/login.ts

```diff
@@ -11,44 +11,31 @@ You should have received a copy of the GNU General Public License along with @p0
 import { OIDC_HEADERS } from "../../common/auth/oidc";
 import { urlEncode, validateResponse } from "../../common/fetch";
 import { print2 } from "../../drivers/stdio";
-import { AuthorizeResponse, TokenResponse } from "../../types/oidc";
+import { AuthorizeResponse, OidcLoginStepHelpers } from "../../types/oidc";
 import { OrgData } from "../../types/org";
-import { sleep, throwAssertNever } from "../../util";
+import { sleep } from "../../util";
 import { capitalize } from "lodash";
 import open from "open";
 
-const DEVICE_GRANT_TYPE = "urn:ietf:params:oauth:grant-type:device_code";
+export const DEVICE_GRANT_TYPE = "urn:ietf:params:oauth:grant-type:device_code";
 
 export const validateProviderDomain = (org: OrgData) => {
   if (!org.providerDomain) throw "Login requires a configured provider domain.";
 };
 
 /** Executes the first step of a device-authorization grant flow */
 // cf. https://developer.okta.com/docs/guides/device-authorization-grant/main/
-const authorize = async (org: OrgData, scope: string) => {
-  if (org.providerType === undefined) {
-    throw "Login requires a configured provider type.";
-  }
-  const init = {
-    method: "POST",
-    headers: OIDC_HEADERS,
-    body: urlEncode({
-      client_id: org.clientId,
-      scope,
-    }),
-  };
-  validateProviderDomain(org);
-  // This is the "org" authorization server; the okta.apps.* scopes are not
-  // available with custom authorization servers
-  const url =
-    org.providerType === "okta"
-      ? `https:${org.providerDomain}/oauth2/v1/device/authorize`
-      : org.providerType === "ping"
-        ? `https://${org.providerDomain}/${org.environmentId}/as/device_authorization`
-        : throwAssertNever(org.providerType);
+export const authorize = async <T>(
+  request: {
+    url: string;
+    init: RequestInit;
+  },
+  validateResponse?: (response: Response) => Promise<Response>
```

Comment:
In which situation is this ever undefined? It might be better to force the consumer to explicitly pass an async noop instead of ommitting this, to prevent unintended bugs.